### PR TITLE
Decouple Helm Release with the Operator Release

### PR DESCRIPTION
**Pull request by prudhvigodithi**
### Description
Decouple Helm Release with the Operator Release.

- Update the `release.yaml` to exclude the helm release.
- Onboard new helm release workflow `helm-release.yaml`.
- The `helm-release.yaml` use the [helm/chart-releaser-action](https://github.com/helm/chart-releaser-action) which identifies the chart version and releases incrementally.
- The `helm-release.yaml` runs once the PR is merged to `main` and releases only once there is a change with chart `version`, inside the [charts/ ](https://github.com/opensearch-project/opensearch-k8s-operator/tree/main/charts)folder.

The expectation from the user is to create a PR with with updates changes, along with modifying the `version` of the chart, once merged the `helm/chart-releaser-action` will identify the new chart version and does a helm release.

Whenever there is an operator version release, the `appVersion` along with `version` has to be updated and thereafter the process is same  `helm/chart-releaser-action` will identify the new chart version and does a helm release.

### Issues Resolved
Coming from https://github.com/opensearch-project/opensearch-k8s-operator/issues/830

### Check List
- [x] Commits are signed per the DCO using --signoff 
- [ ] Unittest added for the new/changed functionality and all unit tests are successful
- [ ] Customer-visible features documented
- [ ] No linter warnings (`make lint`)

If CRDs are changed:
- [ ] CRD YAMLs updated (`make manifests`) and also copied into the helm chart
- [ ] Changes to CRDs documented

Please refer to the [PR guidelines](https://github.com/opensearch-project/opensearch-k8s-operator/blob/main/docs/developing.md#submitting-a-pr) before submitting this pull request.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
Adding @swoehrl-mw and @evheniyt to please take a look, once this is merged we can move forward with this PR https://github.com/opensearch-project/opensearch-k8s-operator/pull/754 to refactor opensearch-cluster helm chart. 

Once this PR is merged the idea is to rebase this PR https://github.com/opensearch-project/opensearch-k8s-operator/pull/754, update the `version` and merge to `main`, then the `helm/chart-releaser-action` will take care of releasing the helm chart.

Thank you 
@getsaurabh02 @peterzhuamazon

**Review by peterzhuamazon — APPROVED** [review]
So basically helm will always be latest HEAD?
While operator official release will be stable ones?

Thanks.

**Comment by prudhvigodithi** [comment]
> So basically helm will always be latest HEAD? While operator official release will be stable ones?
> 
> Thanks.

Thanks Peter, the operator when released creates the docker image and other helm charts (the seperate releases) uses this operator version from `appVersion`. 

It is the same mechanism that we use today for OpenSearch helm charts https://github.com/opensearch-project/helm-charts/pull/569 where the `appVersion` is the OpenSearch version.

**Review by swoehrl-mw — CHANGES_REQUESTED** [review]
I'm fine with doing this for the opensearch-cluster helm chart and can see the advantages of having a faster release cycle.
But for the opensearch-operator chart this is a problem with the current development process: If an operator feature requires changes to the chart (e.g. a new config option or changes to one of the CRDs) these are made in the same PR as the code changes. With this new release process we run a huge risk of releasing features in the chart that are not yet included in the operator code if someone makes a non-related change in the chart and wants to release that by increasing the version. This will only lead to problems.
So IMO the release and version of the operator chart should stay coupled with the operator release.


Also: You at least need to extend the `CONTRIBUTING.md` and `docs/developing.md` to explain the process, that users need to edit the changelog and how and when to increase helm chart versions.

**Comment by prudhvigodithi** [comment]
> With this new release process we run a huge risk of releasing features in the chart that are not yet included in the operator code if someone makes a non-related change in the chart and wants to release that by increasing the version. This will only lead to problems.
> So IMO the release and version of the operator chart should stay coupled with the operator release.

Thanks, @swoehrl-mw. I agree with your point, but I believe the maintainers are responsible for reviewing PRs and Helm chart changes. For example, in this PR https://github.com/opensearch-project/opensearch-k8s-operator/pull/754/files, the change involved the `opensearch-cluster`, and the maintainers should be aware of code modifications made to the `opensearch-operator` Helm chart and decide if the chart version needs to be incremented. 

This PR aims to standardize the Helm release process for both `opensearch-cluster` and `opensearch-operator` Helm charts, which would only require a version increment, and the workflow would handle the release. The timing and coordination of version increments should be managed by the PR authors and maintainers to ensure smooth chart releases. 

In some cases, like a critical/small fix, we may need to release the `opensearch-operator` Helm chart independently, without waiting for the `opensearch-operator` release and this process (by this PR) should be helpful and unblock the chart release.

Thank you 
Adding @getsaurabh02 @salyh @pchmielnik @jochenkressin @dblock @peterzhuamazon

**Comment by prudhvigodithi** [comment]
@swoehrl-mw any other suggestions on how to move forward with this PR https://github.com/opensearch-project/opensearch-k8s-operator/pull/754 without decoupling the releases?

> that is not really an argument for me.

Again we cant release an operator for every small change in the helm chart (and helm chart should not wait until the operator is released), IMO its totally worth to decouple both the 2 products, the operator and the helm chart.

**Comment by prudhvigodithi** [comment]
Following, @swoehrl-mw can you please suggest the next steps here? I want to Decouple Helm Release with the Operator Release because of https://github.com/opensearch-project/opensearch-k8s-operator/pull/872#issuecomment-2364003894 and unblock the PR https://github.com/opensearch-project/opensearch-k8s-operator/pull/754.
Thank you 
@getsaurabh02 @peterzhuamazon

**Comment by swoehrl-mw** [comment]
@prudhvigodithi I would suggest to have the opensearch-cluster helm chart with the release process as you suggest it in this PR and have operator docker + opensearch-operatorhelm chart with the release process as it is currently. That way #754 would be unblocked. Releases for the operator chart would still be slow, but as I said in an earlier comment that chart is tightly coupled with operator features so they should really be released only together.

**Comment by evheniyt** [comment]
@prudhvigodithi for me this sounds reasonable, could we start moving forward with decoupling for opensearch-cluster?

**Comment by prudhvigodithi** [comment]
Sure @evheniyt let me finalize the PR with just decoupling `opensearch-cluster` helm chart.

**Comment by prudhvigodithi** [comment]
Coming from https://github.com/helm/chart-releaser?tab=readme-ov-file#dealing-with-charts-that-have-dependencies leveraged `charts_dir` to release opensearch-operator helm chart with the common operator release and opensearch-cluster on merge to `main` with version increment.

File: .github/workflows/release.yaml
Comment (peterzhuamazon) [inline .github/workflows/release.yaml:26]: Extra spacing?

**Review by peterzhuamazon — APPROVED** [review]

**Review by swoehrl-mw — APPROVED** [review]

**Pull request merged**
